### ISSUE TITLE
Can't query providers with constraints ~> 3.1

### ISSUE THREAD

**Issue by wahlfeld10x** [terraform-providers/terraform-provider-aws#15632]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

*  v3.10.0

### Terraform Configuration Files

```hcl
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.1"
    }
  }
```

### Expected Behavior

```
Initializing provider plugins...
- Finding hashicorp/aws versions matching "~> 3.1"...
- Installing hashicorp/aws v3.10.0...
- Installed hashicorp/aws v3.10.0 (unauthenticated)

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

### Actual Behavior

```
Initializing provider plugins...
- Finding hashicorp/aws versions matching "~> 3.1"...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider hashicorp/aws:
no available releases match the given constraints ~> 3.1
```

This seems to have happened since v3.10.0. We've not made any changes to the provider block.

**Comment by wahlfeld10x** [comment]
Possibly related to this? https://github.com/hashicorp/terraform/issues/26574

**Comment by gdavison** [comment]
Thanks for raising this, @wahlfeld10x. Are you able to try running Terraform with [debug logging enabled](https://www.terraform.io/docs/internals/debugging.html)? It may give us more details on what is happening.

**Comment by gdavison** [comment]
Also include the version of Terraform that you're using, please

**Comment by wahlfeld10x** [comment]
Fixed. I believe we were misinterpreting the docs on version constraints.

Changed this 
`      version = "~> 3.1"`
To this
`      version = "~> 3.10.0"`

And available providers packaged with our deployment image from this
`    versions = ["~> 2.0", "~> 3.9.0"]`
to this
`    versions = ["~> 2.0", "~> 3.9.0", "~> 3.10.0"]`

**Comment by gordonbondon** [comment]
This is the first link that pops up on search when I was looking for a similar issue, so I'll post my workaround.

Basically after upgrading from terraform to 0.12.* to 0.13.4 and later trying to upgrade aws provider from 2.59.0 to 3.11.0 I was receiving similar errors, even though there were no conflicts between constraints:
```
Could not retrieve the list of available versions for provider hashicorp/aws:
no available releases match the given constraints 3.11.0, >= 2.30.0, >=
2.42.*, < 4.0.*, >= 2.42.*, < 4.0.*
```

After I removed my strict version requirement provider was installed with an older version and `(unauthenticated)` message:
```
- Installing hashicorp/aws v2.59.0...
- Installed hashicorp/aws v2.59.0 (unauthenticated)
```

I did not have `provider_installation` or `plugin_cache_dir` configured, but completely deleting contents of old plugin downloads in `~/.terraform.d/plugins/` resolved this issue for me.
